node/impl/full/actor_events.go

```diff
@@ -39,188 +40,320 @@ type ActorEventsAPI struct {
+		}
+	} else if minHeight >= 0 && maxHeight >= 0 {
+		if minHeight > maxHeight {
+			return 0, 0, fmt.Errorf("invalid epoch range: 'to' height (%d) must be after 'from' height (%d)", minHeight, maxHeight)
+		} else if maxHeight-minHeight > maxRange {
+			return 0, 0, fmt.Errorf("invalid epoch range: range between to and 'from' heights is too large (maximum: %d)", maxRange)
+		}
+	}
+	return minHeight, maxHeight, nil
+}
+
+func (a *ActorEventHandler) SubscribeActorEvents(ctx context.Context, evtFilter *types.ActorEventFilter) (<-chan *types.ActorEvent, error) {
 	if a.EventFilterManager == nil {
 		return nil, api.ErrNotSupported
 	}
-
-	params, err := a.parseFilter(&f.Filter)
+	if evtFilter == nil {
+		evtFilter = &types.ActorEventFilter{}
+	}
+	params, err := a.parseFilter(*evtFilter)
 	if err != nil {
 		return nil, err
 	}
 
-	fm, err := a.EventFilterManager.Install(ctx, params.MinHeight, params.MaxHeight, params.TipSetCid, f.Filter.Addresses, f.Filter.Fields, false)
+	tipSetCid, err := params.GetTipSetCid()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tipset cid: %w", err)
+	}
+	fm, err := a.EventFilterManager.Install(ctx, params.MinHeight, params.MaxHeight, tipSetCid, evtFilter.Addresses, evtFilter.Fields, false)
 	if err != nil {
 		return nil, err
 	}
 
-	out := make(chan *types.ActorEvent, 25)
+	// The goal for the code below is to be able to send events on the `out` channel as fast as
+	// possible and not let it get too far behind the rate at which the events are generated.
+	// For historical events we see the rate at which they were generated by looking the height range;
+	// we then make sure that the client can receive them at least twice as fast as they were
+	// generated so they catch up quick enough to receive new events.
+	// For ongoing events we use an exponential moving average of the events per height to make sure
+	// that the client doesn't fall behind.
+	// In both cases we allow a little bit of slack but need to avoid letting the client bloat the
+	// buffer too much.
+	// There is no special handling for reverts, so they will just look like a lot more events per
+	// epoch and the user has to receive them anyway.
+
+	out := make(chan *types.ActorEvent)
 
 	go func() {
 		defer func() {
-			// Tell the caller we're done
+			// tell the caller we're done
 			close(out)
-
-			// Unsubscribe.
 			fm.ClearSubChannel()
-			_ = a.EventFilterManager.Remove(ctx, fm.ID())
+			if err := a.EventFilterManager.Remove(ctx, fm.ID()); err != nil {
+				log.Warnf("failed to remove filter: %s", err)
+			}
 		}()
 
-		if f.Prefill {
-			evs, err := getCollected(ctx, fm)
-			if err != nil {
-				log.Errorf("failed to get collected events: %w", err)
-				return
+		// Handle any historical events that our filter may have picked up -----------------------------
+
+		evs, minEpoch, maxEpoch := getCollected(ctx, fm)
+		if len(evs) > 0 {
+			// must be able to send events at least twice as fast as they were generated
```

Comment:
You know, I'd give them a single block time and that's it. We already have a limit on the number of events they can collect, and that'll keep us from backing up our main channel.

It'll also simplify this code quite a bit.